### PR TITLE
add method flush_all

### PR DESCRIPTION
**Pull request by hy05190134**
support get_stale method

### CODE REVIEW

**Comment by hy05190134** [comment]
@agentzh  pr for todolist

**Comment by agentzh** [comment]
@hy05190134 I think the `get()` method already returns any stale item. That's not enough for your use cases?

File: lib/resty/lrucache.lua [hunk 1]
```diff
@@ -161,6 +161,28 @@ function _M.get(self, key)
 end
 
 
+function _M.get_stale(self, key)
```
Comment (agentzh) [inline lib/resty/lrucache.lua]: The `get()` method already returns the stale data. I don't think this adds anything new.

File: lib/resty/lrucache.lua [hunk 2]
```diff
@@ -223,5 +245,25 @@ function _M.set(self, key, value, ttl)
...
 end
 
+function _M.flush_all(self)
+    local cache_queue = self.cache_queue
+    local free_queue = self.free_queue
+    local node2key = self.node2key
+    local hasht = self.hasht
+    local key2node = self.key2node
+    while not queue_is_empty(cache_queue)
+    do
```
Comment (agentzh) [inline lib/resty/lrucache.lua]: Style: we usually put `do` in the previous line containing `while`.

File: lib/resty/lrucache.lua [hunk 3]
```diff
@@ -223,5 +245,25 @@ function _M.set(self, key, value, ttl)
...
+        node = queue_last(cache_queue)
+
+        local key = node2key[ptr2num(node)]
+        if key then
+            hasht[key] = nil
+            key2node[key] = nil
+        end
+
```
Comment (agentzh) [inline lib/resty/lrucache.lua]: You did not clear the `node2key` table?

Comment (agentzh) [inline lib/resty/lrucache.lua]: Also, I think it's much more efficient to use `table.clear()` to clear those Lua tables instead of manually deleting each individual keys in this while loop.

File: lib/resty/lrucache.lua (hunk 1 above)
Comment (hy05190134) [inline lib/resty/lrucache.lua]: so you mean it is not necessary to add the get_stale method, so you can just modify the readme   description

File: lib/resty/lrucache.lua (hunk 3 above)
Comment (hy05190134) [inline lib/resty/lrucache.lua]: sure, will you agree to merge if I add the method  flush_all?

Comment (agentzh) [inline lib/resty/lrucache.lua]: flush all looks useful.

Comment (hy05190134) [inline lib/resty/lrucache.lua]: but how to handle free_queue and cache_queue, will it cause memory_leak if I just use "self.cache_queue = queue_init() and self.free_queue = queue_init(size), or also I should use while to do in the loop"

File: lib/resty/lrucache.lua [hunk 4]
```diff
@@ -10,7 +10,7 @@ local ngx_now = ngx.now
 local uintptr_t = ffi.typeof("uintptr_t")
 local setmetatable = setmetatable
 local tonumber = tonumber
-
+local tb_clear = table.clear
```
Comment (agentzh) [inline lib/resty/lrucache.lua]: We need 2 blank lines here.

File: lib/resty/lrucache.lua [hunk 5]
```diff
@@ -10,7 +10,7 @@ local ngx_now = ngx.now
...
```
Comment (agentzh) [inline lib/resty/lrucache.lua]: I believe the right syntax is `require "table.clear"`. Also, we should keep compatibility with LuaJIT 2.0 which lacks `table.clear` by providing our own stub implementation in this case.

File: t/sanity.t [hunk 6]
```diff
@@ -247,3 +247,34 @@ dog: 33
 
 --- no_error_log
 [error]
+
+
```
Comment (agentzh) [inline t/sanity.t]: We need 3 blank lines here. Usually we use the `reindex` script to automatically re-format `.t` files.

File: lib/resty/lrucache.lua (hunk 5 above)
Comment (agentzh) [inline lib/resty/lrucache.lua]: This error also makes the whole test suite failing. See the Travis CI report for this PR.

Comment (hy05190134) [inline lib/resty/lrucache.lua]: I know you mean I should just implement the function of table.clear when user use luajit 2.0, but how to check whether use luajit 2.0 or luajit 2.1 ?

**Comment by agentzh** [comment]
@hy05190134 We have a similar example in lua-resty-core: https://github.com/openresty/lua-resty-core/blob/master/lib/resty/core/base.lua#L30

**Comment by hy05190134** [comment]
@agentzh  whether to merge now?

**Comment by agentzh** [comment]
@hy05190134 I'll look into this as soon as I can manage.

**Comment by thibaultcha** [comment]
This would be nice. I believe we would need the same for the `pureffi` implementation as well though?

**Comment by agentzh** [comment]
@thibaultcha Agreed.

File: lib/resty/lrucache.lua [hunk 7]
```diff
@@ -223,5 +238,17 @@ function _M.set(self, key, value, ttl)
     end
 end
```
Comment (agentzh) [inline lib/resty/lrucache.lua]: Style: we use 2 blank lines to separate Lua function definitions consistently in this Lua library.

File: lib/resty/lrucache.lua [hunk 8]
```diff
@@ -223,5 +238,17 @@ function _M.set(self, key, value, ttl)
...
 
+function _M.flush_all(self)
+    tb_clear(self.hasht)
+    tb_clear(self.node2key)
+    tb_clear(self.key2node)
+    local cache_queue = self.cache_queue
```
Comment (agentzh) [inline lib/resty/lrucache.lua]: Better add a blank line before this line to make the code less crowded.

File: lib/resty/lrucache.lua [hunk 9]
```diff
@@ -223,5 +238,17 @@ function _M.set(self, key, value, ttl)
...
+    local free_queue = self.free_queue
+    while not queue_is_empty(cache_queue) do
```
Comment (agentzh) [inline lib/resty/lrucache.lua]: Ditto.

**Comment by thibaultcha** [comment]
@hy05190134 I came up with an FFI implementation in this commit if you want to incorporate it in this PR:

https://github.com/thibaultcha/lua-resty-lrucache/commit/6775038c7354ab695829b09ed4205efd500f14bb

**Comment by hy05190134** [comment]
@thibaultcha pleasure to add method supporting ffi, how should I do to merge your commit?

**Comment by hy05190134** [comment]
@agentzh  I have modified the style

**Comment by hy05190134** [comment]
@thibaultcha  you have already merged my commit into your commit, I see

**Comment by thibaultcha** [comment]
@hy05190134 Feel free to just copy-paste it :)

**Comment by hy05190134** [comment]
@thibaultcha  I have copy-paste some of your code to support ffi

**Comment by hy05190134** [comment]
@agentzh  @thibaultcha  is it OK to merge?

File: lib/resty/lrucache.lua [hunk 10]
```diff
@@ -224,4 +239,19 @@ function _M.set(self, key, value, ttl)
...
+    local cache_queue = self.cache_queue
+    local free_queue = self.free_queue
+
+    while not queue_is_empty(cache_queue) do
+        local node = cache_queue[0].prev
+        queue_remove(node)
+        queue_insert_tail(free_queue, node)
+    end
+end
+
```
Comment (thibaultcha) [inline lib/resty/lrucache.lua]: style: we need 2 lines jumps here

File: lib/resty/lrucache.lua [hunk 11]
```diff
@@ -11,6 +11,21 @@ local uintptr_t = ffi.typeof("uintptr_t")
 local setmetatable = setmetatable
 local tonumber = tonumber
 
+if string.find(jit.version, " 2.0") then
+    ngx.log(ngx.ALERT, "use of lua-resty-lrucache with LuaJIT 2.0 is ",
+    "not recommended; use LuaJIT 2.1+ instead")
```
Comment (thibaultcha) [inline lib/resty/lrucache.lua]: style: I believe the preferred form is to align this line with the 2nd argument, like so:
```lua
ngx.log(ngx.ALERT, "use of lua-resty-lrucache with LuaJIT 2.0 is ",
                   "not recommended; use LuaJIT 2.1+ instead")
```

File: t/sanity.t [hunk 12]
```diff
@@ -247,3 +247,70 @@ dog: 33
...
+
+=== TEST 7: flush value
+--- http_config eval: $::HttpConfig
+--- config
+    location = /t {
+        content_by_lua '
```
Comment (thibaultcha) [inline t/sanity.t]: New tests should use the `*_by_lua_block` form for the ngx_lua directives (just like the `pureffi` test from my commit). Maybe, both tests (pure Lua, and pureffi) should be identical too? The only difference should be the required module imho.

File: lib/resty/lrucache.lua [hunk 13]
```diff
@@ -11,6 +11,21 @@ local uintptr_t = ffi.typeof("uintptr_t")
 local setmetatable = setmetatable
 local tonumber = tonumber
 
+if string.find(jit.version, " 2.0") then
```
Comment (thibaultcha) [inline lib/resty/lrucache.lua]: style: there _might_ be need for 2 line jumps here as well...

**Comment by hy05190134** [comment]
@thibaultcha  thanks for your suggestion, I have already modified, do you have authority to merge?

File: lib/resty/lrucache.lua (hunk 10 above)
Comment (hy05190134) [reply lib/resty/lrucache.lua]: agree

File: lib/resty/lrucache.lua (hunk 11 above)
Comment (hy05190134) [reply lib/resty/lrucache.lua]: agree

File: lib/resty/lrucache.lua (hunk 13 above)
Comment (hy05190134) [reply lib/resty/lrucache.lua]: agree

**Comment by thibaultcha** [comment]
I do not. Let's wait for @agentzh to take a look at this.

File: lib/resty/lrucache.lua [hunk 14]
```diff
@@ -12,6 +12,22 @@ local setmetatable = setmetatable
 local tonumber = tonumber
 
 
+if string.find(jit.version, " 2.0") then
```
Comment (agentzh) [inline lib/resty/lrucache.lua]: Will you please port this change over?

https://github.com/openresty/lua-resty-core/commit/9e6d1effc

File: t/sanity.t [hunk 15]
```diff
@@ -247,3 +247,73 @@ dog: 33
...
+key 1: nil
+key 2: nil
+key 3: nil
+
+--- no_error_log
+[error]
+
+
+
+=== TEST 8: flush_all() deletes all keys [pureffi]
```
Comment (agentzh) [inline t/sanity.t]: Will you please move this test case over into `t/pureffi/sanity.t` instead? It does not belong here :) Thanks!

**Comment by hy05190134** [comment]
@agentzh  please review again

File: t/sanity.t [hunk 16]
```diff
@@ -247,3 +247,38 @@ dog: 33
 
 --- no_error_log
 [error]
+
+
+
+=== TEST 7: flush_all() deletes all keys
```
Comment (agentzh) [inline t/sanity.t]: Will you add a test case for flushing an empty cache store?

File: t/pureffi/sanity.t [hunk 17]
```diff
@@ -388,3 +388,38 @@ dog: 33
 
 --- no_error_log
 [error]
+
+
+
+=== TEST 13: flush_all() deletes all keys [pureffi]
```
Comment (agentzh) [inline t/pureffi/sanity.t]: Like below.

**Comment by hy05190134** [comment]
@agentzh  is it OK to merge?

**Comment by agentzh** [comment]
@hy05190134 We'll have another look this week. Thanks!

File: lib/resty/lrucache/pureffi.lua [hunk 18]
```diff
@@ -531,4 +531,18 @@ function _M.set(self, key, value, ttl)
 end
 
 
+function _M.flush_all(self)
+    local cache_queue = self.cache_queue
+    local key_v = self.key_v
+
+    local node = queue_head(cache_queue)
+
+    while node ~= self.cache_queue do
```
Comment (thibaultcha) [inline lib/resty/lrucache/pureffi.lua]: Here, we could use `cache_queue` instead of `self.cache_queue`.

File: lib/resty/lrucache.lua (hunk 10 above)
Comment (yangshuxin) [reply lib/resty/lrucache.lua]: The time-complexity of the while-loop is O(n), where the n is the number of node in the cache-queue. I'm wondering if we can directly splice cache_queue with the free_queue, reducing the complexity down to O(1).

File: lib/resty/lrucache.lua [hunk 19]
```diff
@@ -224,4 +240,28 @@ function _M.set(self, key, value, ttl)
...
+        local free_last = free_queue[0].prev
+        local cache_first = cache_queue[0]
+        local cache_last = cache_first.prev
+
+        free_last.next = cache_first
+        cache_first.prev = free_last
+
+        cache_last.next = free_queue
+        free_first.prev = cache_last
+    end
```
Comment (thibaultcha) [inline lib/resty/lrucache.lua]: With this new change, I went to the new `flush_all()` tests and I added a test case where I filled the LRU (max capacity), then called `flush_all()`, and then tried to set some more keys once again, but what ends up happening is that the LRU becomes unusable once `flush_all()` has been called.

I highly recommend we add one such test case (reusing a filled or half-filled LRU after `flush_all()`), and I also suggest that we revisit this logic, which appears flawed to me?

Comment (agentzh) [reply lib/resty/lrucache.lua]: @thibaultcha Better share your test cases that expose the problem with @hy05190134 ? Thanks!

Comment (thibaultcha) [reply lib/resty/lrucache.lua]: Will do, but I left it on my work laptop for now 😅

Comment (thibaultcha) [reply lib/resty/lrucache.lua]: Here it is:

```
=== TEST 9: flush_all() flush full cache store and allows subsequent reuse
--- http_config eval: $::HttpConfig
--- config
    location = /t {
        content_by_lua_block {
            local N = 100

            local lrucache = require "resty.lrucache"
            local c = lrucache.new(N)

            for i = 1, N do
                c:set("key " .. i, true)
            end

            c:flush_all()

            c:set("new_key", true)

            ngx.say("value set after flush_all: ", c:get("new_key"))
        }
    }
--- request
    GET /t
--- response_body
value set after flush_all: true

--- no_error_log
[error]
```

I suggest we potentially write the same test with half-filled LRUs, and an empty one as well.

**Comment by hy05190134** [comment]
@thibaultcha  I will revisit the logic, thanks for your test case

**Comment by hy05190134** [comment]
@thibaultcha I have modified some code for none-ffi, could you help add some code for ffi?

File: lib/resty/lrucache.lua [hunk 20]
```diff
@@ -260,6 +260,8 @@ function _M.flush_all(self)
 
         cache_last.next = free_queue
         free_first.prev = cache_last
+
+        self.cache_queue = queue_init()
```
Comment (agentzh) [inline lib/resty/lrucache.lua]: You know that `queue_init()` allocates a new queue? Then all your queue recycling logic above is completely useless.

**Comment by hy05190134** [comment]
@agentzh  please review again

**Comment by agentzh** [comment]
@thibaultcha @yangshuxin Will you please have another look at this new version? Thanks!

**Comment by hy05190134** [comment]
@thibaultcha

**Comment by thibaultcha** [comment]
I played with this locally yesterday, and everything seemed ok. Ran Valgrind and naive memory leak test modes successfully. I would maybe add a few more complex test cases, but that's mostly it...

File: lib/resty/lrucache.lua [hunk 21]
```diff
@@ -224,4 +240,31 @@ function _M.set(self, key, value, ttl)
...
+
+        free_last.next = cache_first
+        cache_first.prev = free_last
+
+        cache_last.next = free_queue
+        free_first.prev = cache_last
+
+        cache_queue[0].next = cache_queue
+        cache_queue[0].prev = cache_queue
+    end
```
Comment (thibaultcha) [inline lib/resty/lrucache.lua]: To be slightly more readable 9and save a few index lookups), the splicing logic can be rewritten as:
```lua
local free_head = free_queue[0]
local free_last = free_head.prev

local cache_head = cache_queue[0]
local cache_first = cache_head.next
local cache_last = cache_head.prev

free_last.next = cache_first
cache_first.prev = free_last

cache_last.next = free_head
free_head.prev = cache_last

cache_head.next = cache_queue
cache_head.prev = cache_queue
```

**Comment by agentzh** [comment]
@hy05190134 Thanks for the patch! I've moved your tests from `sanity.t` to a separate `flush-all.t` test file. Also I added more complicated logic and checks to your tests as per @thibaultcha's suggestion. Finally, I added some docs for this new method.
